DMFT/src/main/java/com/gov/rsi/DMFT/controllers/ReportController.java

```diff
@@ -14,51 +21,123 @@
+	private static final String LICENSE_NUMBER_PATH = "$..item[?(@.linkId==\"{0}\")]";
+	// The linkId of the "item" element holding the license number
+	private static final String LICENSE_NUMBER_ELEMENT = "patientDriversLicenseNumber";
+	
+	private static Logger log = LoggerFactory.getLogger(ReportController.class);
 	
 	@Autowired
 	private ReportRepository repository;
 
+	/**
+	 * Fetches the oldest report with NEW status from the queue
+	 * @return the DMER Json report
+	 */
 	@RequestMapping(value = "/", method = RequestMethod.GET)
-	public List<Report> getAllReports() {
+	public ResponseEntity<?> getNextNewReport() {
 		
-		return repository.findAll();
-	  
-	}
-	
-	@RequestMapping(value = "/{id}", method = RequestMethod.GET)
-	public Report getReportById(@PathVariable("id") ObjectId id) {
+		List<Report> reports = repository.findFirst1ByStatusOrderByTimeNewAsc(Report.Status.NEW);
+
+		if (reports.size() == 0) {
+			// No new ones left in queue
+			return responseNoContent();
+		}
 		
-		return repository.findBy_id(id);
-	  
-	}
-	
-	@RequestMapping(value = "/{id}", method = RequestMethod.PUT)
-	public Report modifyReportById(@PathVariable("id") ObjectId id, 
-			@Valid @RequestBody Object report) {
+		// Switch status to In Process
+		Report nextOne = reports.get(0);
+		nextOne.setStatus(Report.Status.INPROCESS);
+		repository.save(nextOne);
 		
-		Report driversReport = new Report(id, "In-Progress", report);
-		repository.save(driversReport);
-		return driversReport;
+		return responseOkWithBody(nextOne.getJson());
 	}
 	
+	/**
+	 * Inserts a DMER JSON report into the queue
+	 * @param dmerJson
+	 * @return
+	 */
 	@RequestMapping(value = "/", method = RequestMethod.POST)
-	public Report createReport(@Valid @RequestBody Object report) {
+	public ResponseEntity<?> createReport(@RequestBody String dmerJson) {
 		
-		Report driversReport = new Report(ObjectId.get(), "New", report);
-		repository.save(driversReport);
-		return driversReport;		
-	}
-	
-	@RequestMapping(value = "/{id}", method = RequestMethod.DELETE)
-	public void deleteReport(@PathVariable ObjectId id) {
+		String licenseNumber = null;
 		
-		repository.delete(repository.findBy_id(id));
+		// Locate the driver's license within the DMER Json text 
+		try {
+			String query = MessageFormat.format(LICENSE_NUMBER_PATH, LICENSE_NUMBER_ELEMENT);		
+			licenseNumber = JsonPath.read(
+					JsonPath.read(dmerJson, query)
+					.toString(), "$[0].answer[0].valueDecimal").toString();
+		}
+		catch (PathNotFoundException e) {
+			// Treat same as if path found but element is empty
+		}
 		
-	}	
+		// TODO: Need to apply any other criteria for a valid DL number (e.g. 7 digits only ) ?
+		if (licenseNumber == null || licenseNumber.length() == 0) {
+			log.warn("DMER JSON: Unable to locate DL");
+			return responseBadRequest();
+		}
+		
+		// Look for records with matching DL
+		List<Report> reports = repository.findByLicenseNumber(licenseNumber);

```

Comment:
Have we created an index in MongoDB for this property?

Reply:
Yes, the @Index annotation in the Dmer class does this.